---
title: "Crosswalking Global Coral Microbiome Data (GCMP) into the Genomic Observatories Metadatabase (GEOME)"
output: html_notebook
---

# Introduction

This notebook documents the code used to translate the Global Coral Microbiome Project's (GCMP) collated metadata (provided by Jesse Zaneveld) into a format that GEOME can ingest. We are hoping to share the GCMP metadata on GEOME and for this to be a use-case to help improve GEOME's ability to store microbiome metadata.

The metadata provided by Jesse are in `GCMP_EMP_map_r29.txt`. The template to which we want to write the data is in `Global Coral Microbiome Project.xlsx`

# Setup

```{r setup}
library(tidyverse)
library(lubridate)
library(xlsx)

```

# Importing the data

```{r}

y_or_n <- function(x){
  x == "y"
}

df <- read_tsv(file = "GCMP_EMP_map_r29.txt", na = c("","NA","na","Not Applicable"),
              col_types = cols(
                date = col_date(format = "%m/%d/%y"),
                depth = "n",
                temperature = "n",
                surface_temperature = "n",
                photosynthetically_active_radiation = "n",
                colony_height_m = "n",
                colony_width1 = "n",
                colony_width2 = "n",
                tissue_loss_percent = "n",
                disease_percent = "n",
                cyanobacteria_percent = "n",
                turf_contact_percent = "n",
                n_algal_contacts = "i",
                n_macroalgal_contacts = "i",
                cca_contact_percent = "n",
                TAXON_ID = "n",
                Colony.maximum.diameter = "n",
                Corallite.width.maximum = "n",
                Corallite.width.minimum = "n",
                Depth.lower = "n",
                Depth.upper = "n",
                Eastern.most.range.edge = "n",
                Genus.fossil.age = "n",
                Northern.most.range.edge = "n",
                Oocyte.size.at.maturity = "n",
                Propagule.size.on.release.max = "n",
                Propagule.size.on.release.mean = "n",
                Range.size = "n",
                Skeletal.density = "n",
                Southern.most.range.edge = "n",
                Species.age.phylogeny = "n",
                Western.most.range.edge = "n",
                prop_Colony_maximum_GCMP_recorded = "n",
                prop_Colony_maximum_diameter_universal = "n",
                prop_Colony_maximum_diameter = "n",
                Colony_maximum_diameter = "n",
                Colony_maximum_GCMP_recorded = "n",
                Colony_maximum_diameter_universal = "n",
                visibility = "n"
              )) 
          #this optionally changes all columns that start with "binary" to logical (T/F) instead of y/n
          #mutate(across(starts_with("binary"), function(x){x == "y"}
          #              )) 
          # you may want to continue cleaning the data in this import pipeline statement, 
          # like changing other y/n columns to logical  
  
# Check column class
summary=sapply(df, class)
write.csv(summary, "summary-table-class.csv")
```

# Clean the Data


# Split the data into GEOME tables

Now you'll need to split the columns into the four tables that GEOME uses: events, samples, tissues, and diagnostics  

You can follow what I specified in GCMP_EMP_map_r29_EDC.xlsx (green = events, blue = samples, yellow = tissues, orange = diagnostics)

Still, some decisions need to be made about what gets included in GEOME or not. XXXXX means I think it 
shouldn't be included, ???? means I don't know you should ask or find a darwinCore term, and then I made no decisions
past family (column CU)

```{r}
Event <- df %>% select(collection_id,
                          date,
                          reef_name,
                          collection_time,
                          latitude,
                          longitude,
                          Range.size,
                          reef_type,
                          site_name,
                          geographic_area,
                          ocean_area,
                          ocean,
                          ocean_superregion,
                          country,
                          collected_by)
Sample <- df %>% select(sample_name_backup,
                          sample_type_EMP,
                          sample_name_first_demultiplexing,
                          physical_sample_location,
                          physical_sample_location,
                          local_colony_name,
                          local_sample_id,
                          colony_name,
                          host_scientific_name,
                          host_genus_id,
                          host_species_id,
                          biological_sample_notes,
                          Huang_Roy_tree_name,
                          oz_disease_mean,
                          oz_disease_sd,
                          host_clade_sensu_fukami_numeric,
                          host_clade_sensu_fukami,
                          NCBI_inherited_blast_name,
                          full_taxonomy_string_with_clade,
                          taxonomy_string_to_clade,
                          taxonomy_string_to_family,
                          taxonomy_string_to_order,
                          taxonomy_string_to_subclass,
                          taxonomy_string_to_class,
                          complex_robust,
                          family,
                          outgroup,
                          common_name,
                          synonyms,
                          species_notes,
                          Description)
Sample_or_Event <- df %>% select(substrate, depth)
Tissue <- df %>% select(SampleID,
                          compartment_replicates,
                          sample_type,
                          tissue_compartment,
                          BiologicalMatter)
Diagnostics <- df %>% select(temperature,
                              surface_temperature,
                              salinity,
                              oxygen,
                              photosynthetically_active_radiation,
                              PAR_method,
                              contact_description,
                              colony_height_m,
                              colony_width1,
                              colony_width2,
                              coral_color,
                              coral_color_additional,
                              binary_tissue_loss,
                              tissue_loss_percent,
                              disease_name,
                              binary_disease,
                              disease_percent,
                              cyanobacteria_contact,
                              cyanobacteria_percent,
                              cyanobacteria_overgrown,
                              binary_turf_contact,
                              turf_contact_percent,
                              turf_overgrown,
                              binary_algal_contact,
                              n_algal_contacts,
                              algal_contact_types,
                              binary_macroalgal_contact,
                              n_macroalgal_contacts,
                              macroalgal_contact_types,
                              binary_CCA_contact,
                              cca_contact_percent,
                              cca_overgrown,
                              seagrass_contact,
                              parrotfish_bites,
                              invertebrate_contact_types,
                              coral_contact_types,
                              sediment_contact,
                              corallite_type,
                              calcifies,
                              functional_group_sensu_darling,
                              Abundance.GBR,
                              Abundance.world,
                              Coloniality,
                              Colony.maximum.diameter,
                              Corallite.width.maximum,
                              Corallite.width.minimum,
                              Depth.lower,
                              Depth.upper,
                              Genus.fossil.age,
                              Growth.form.typical,
                              IUCN.Red.List.category,
                              Mode.of.larval.development,
                              Oocyte.size.at.maturity,
                              Propagule.size.on.release.max,
                              Propagule.size.on.release.mean,
                              Sexual.system,
                              Skeletal.density,
                              Species.age.phylogeny,
                              Substrate.attachment,
                              Water.clarity.preference,
                              Wave.exposure.preference,
                              Zooxanthellate,
                              Colony_maximum_GCMP_recorded,
                              shelf_location,
                              visibility,
                              visibility_method)



```

# Rename and convert

Now you need to rename and/or convert any data into GEOME format

```{r}
events <- events %>% rename(SampleID = tissueID, 
                            collection_id = locationID,
                            date = yearCollectedANDmonthCollectedANDdayCollected,# 
                            reef_name =locality, 
                            collection_time = timeOfDay,
                            latitude = decimalLatitude,
                            longitude = decimalLongitude,
                            substrate = substratum,
                            depth =	minimumDepthInMeters,#
                            temperature = temperatureInDegreesCelsius,
                            surface_temperature = surfaceTemperatureinDegreesCelsius,
                            salinity = salinity,#
                            photosynthetically_active_radiation	= PARinUnits,
                            PAR_method = PARMethod,
                            physical_sample_location =	principalInvestigator,#
                            sample_type = tissueType,
                            local_colony_name = verbatimLabel, 
                            contact_description = contactDescription,
                            local_sample_id = individualID, 
                            colony_sample_name = materialSampleID,
                            host_scientific_name = scientificName,
                            host_genus_id = genus,
                            host_species_id = specificEpithet,
                            colony_height_cm = colonyHeightInMeters,
                            colony_width1 = colonyWidth1InMeters,
                            colony_width2 =colonyWidth2InMeters,
                            coral_color = colorWatch,
                            coral_color_additional = colorWatch2,
                            binary_tissue_loss = tissueLoss,
                            tissue_loss_percent = tissueLossPercent,
                            disease_name = diseaseName, 
                            binary_disease = disease,
                            disease_percent = diseasePercent,
                            cyanobacteria_contact = cyanobacteriaContact,
                            cyanobacteria_percent = cyanobacteriaContactPercent,
                            cyanobacteria_overgrown = cyanobacteriaOvergrown,
                            binary_turf_contact = turfContact,
                            turf_contact_percent = turfContactPercent,
                            turf_overgrown = turfOvergrown,
                            binary_algal_contact = algalContact,
                            n_algal_contacts = algalcontactsNumber,
                            algal_contact_types = algalContactTypes,
                            binary_macroalgal_contact = macroalgalContact,
                            n_macroalgal_contacts = macroalgalContactsNumber,
                            macroalgal_contact_types = macroalgalContactTypes,
                            binary_CCA_contact = CCAContact,
                            cca_contact_percent = CCAContactPercent,
                            cca_overgrown = CCAOvergrown,
                            seagrass_contact = seagrassContact,
                            parrotfish_bites = parrotfishBites,
                            invertebrate_contact_types = invertContactTypes,
                            coral_contact_types = coralContactTypes,
                            Sediment_contact = sedimentContact,
                            biological_sample_notes = organismRemarks,
                            field_host_clade_sensu_fukami = higherClassificationANDnamePublishedIn,
                            TAXON_ID = taxonID,
                            full_taxonomy_string_with_clade = taxonomyPhylaANDtaxonomyClassANDtaxonomyOrderANDtaxonomyCladeANDgenusANDspecificEpithet,                           taxonomy_string_to_clade = taxonomyPhylaANDtaxonomyClassANDtaxonomyOrderANDtaxonomyClade,
                            taxonomy_string_to_family = taxonomyPhylaANDtaxonomyClassANDtaxonomyOrderANDtaxonomyFamily,
                            taxonomy_string_to_order = taxonomyPhylaANDtaxonomyClassANDtaxonomyOrder,
                            taxonomy_string_to_subclass = taxonomyPhylaANDtaxonomyClassANDtaxonomySubclas,
                            
                            ) %>% mutate(collectionDay = day(date),
                                                                     collectionMonth = month(date),
                                                                     collectionYear = year(date)) %>%
              select(-date)



```

# Deduplicate

Here is where you'll make the data more efficient, into [First Normal Form](https://en.wikipedia.org/wiki/First_normal_form), 
by removing duplicate entries in the events and samples tables.

Currently consulting with John Deck about moving substratum and depth to the Sample table, rather than events...

```{r}
dim(events)
events <- events %>% distinct()

dim(events)

```


# Split into Expeditions

There were 11 or 14 GCMP expeditions, and we need to split these concatenated data back into their constituent expeditions.

```{r}
E1 <- df %>% mutate(expedition = str_match(SampleID,"10895\\.(E\\d+)")[,2], .before = sample_name_backup) %>% 
        filter(expedition == "E1")
  

```

# Validate

You'll want to use dplyr left_join() here and some sort of diff function to see if you can exactly reconstruct the dataset from these de-duplicated tables



# Write to Excel Template

You'll want to use the `write.xlsx()` function in the xlsx package to write each table to the `Global Coral Microbiome Project.xlsx` GEOME template in the working directory.