---
title: "Crosswalking Global Coral Microbiome Data (GCMP) into the Genomic Observatories Metadatabase (GEOME)"
output: html_notebook
---

# Introduction

This notebook documents the code used to translate the Global Coral Microbiome Project's (GCMP) collated metadata (provided by Jesse Zaneveld) into a format that GEOME can ingest. We are hoping to share the GCMP metadata on GEOME and for this to be a use-case to help improve GEOME's ability to store microbiome metadata.

The metadata provided by Jesse are in `GCMP_EMP_map_r29.txt`. The template to which we want to write the data is in `Global Coral Microbiome Project.xlsx`

# Setup

```{r setup}
library(tidyverse)
library(lubridate)
library(xlsx)

```

# Importing the data

```{r}

y_or_n <- function(x){
  x == "y"
}

df <- read_tsv(file = "GCMP_EMP_map_r29.txt", na = c("","NA","na","Not Applicable"),
              col_types = cols(
                date = col_date(format = "%m/%d/%y"),
                depth = "n",
                temperature = "n",
                surface_temperature = "n",
                photosynthetically_active_radiation = "n",
                colony_height_m = "n",
                colony_width1 = "n",
                colony_width2 = "n",
                tissue_loss_percent = "n",
                disease_percent = "n",
                cyanobacteria_percent = "n",
                turf_contact_percent = "n",
                n_algal_contacts = "i",
                n_macroalgal_contacts = "i",
                cca_contact_percent = "n",
                TAXON_ID = "n",
                Colony.maximum.diameter = "n",
                Corallite.width.maximum = "n",
                Corallite.width.minimum = "n",
                Depth.lower = "n",
                Depth.upper = "n",
                Eastern.most.range.edge = "n",
                Genus.fossil.age = "n",
                Northern.most.range.edge = "n",
                Oocyte.size.at.maturity = "n",
                Propagule.size.on.release.max = "n",
                Propagule.size.on.release.mean = "n",
                Range.size = "n",
                Skeletal.density = "n",
                Southern.most.range.edge = "n",
                Species.age.phylogeny = "n",
                Western.most.range.edge = "n",
                prop_Colony_maximum_GCMP_recorded = "n",
                prop_Colony_maximum_diameter_universal = "n",
                prop_Colony_maximum_diameter = "n",
                Colony_maximum_diameter = "n",
                Colony_maximum_GCMP_recorded = "n",
                Colony_maximum_diameter_universal = "n",
                visibility = "n"
              )) 
          #this optionally changes all columns that start with "binary" to logical (T/F) instead of y/n
          #mutate(across(starts_with("binary"), function(x){x == "y"}
          #              )) 
          # you may want to continue cleaning the data in this import pipeline statement, 
          # like changing other y/n columns to logical  
  
# Check column class
summary=sapply(df, class)
write.csv(summary, "summary-table-class.csv")
```

# Clean the Data


# Split the data into GEOME tables

Now you'll need to split the columns into the four tables that GEOME uses: events, samples, tissues, and diagnostics  

You can follow what I specified in GCMP_EMP_map_r29_EDC.xlsx (green = events, blue = samples, yellow = tissues, orange = diagnostics)

Still, some decisions need to be made about what gets included in GEOME or not. XXXXX means I think it 
shouldn't be included, ???? means I don't know you should ask or find a darwinCore term, and then I made no decisions
past family (column CU)



# Rename and convert

Now you need to rename and/or convert any data into GEOME format

```{r}
Event <- Event %>% rename(locationID = collection_id,
                          timeOfDay = collection_time,
                          decimalLatitude = latitude,
                          decimalLongitude = longitude,
                          substratum = substrate, 
                          maximumDepthInMeters = depth,
                          rangeSizeInSquareKilometers = Range.size,###
                          fieldNotes = reef_type,
                          cardinalLocation = geographic_area,
                          waterBody = ocean_area,
                          oceanBasin = ocean,
                          oceanSuperregion = ocean_superregion,
                          country = country,
                          collectorList = collected_by,
                          principalInvestigator = relevant_collaborators) %>%
                          unite(locality, reef_name, site_name) %>% # locality appeared twice for reef_name and site_name, so I combined the two columns, delimited by an underscore
                          mutate(yearCollected = year(date),
                            monthCollected = month(date),
                            dayCollected = day(date)) %>% # Separate date into year, month, day in that order
                            




Sample <- df %>% select(basisOfRecord = sample_type_EMP, 
                          institutionCode = physical_sample_location,
                          verbatimLabel = local_colony_name,
                          individualID = local_sample_id,
                          materialSampleID = colony_name,
                          scientificName = host_scientific_name,
                          genus = host_genus_id,
                          specificEpithet = host_species_id,
                          organismRemarks = biological_sample_notes,
                          Huang_Roy_tree_name = Huang_Roy_tree_name,
                          higherClassification = host_clade_sensu_fukami,
                          taxonID = TAXON_ID,
                          complex_robust = complex_robust,
                          full_taxonomy_string_with_clade = full_taxonomy_string_with_clade, 
                          taxonomy_string_to_clade = taxonomy_string_to_clade,
                          taxonomy_string_to_family = taxonomy_string_to_family,
                          taxonomy_string_to_order = taxonomy_string_to_order,
                          taxonomy_string_to_subclass = taxonomy_string_to_subclass,
                          taxonomy_string_to_class = taxonomy_string_to_class,
                          family = family,
                          colloquialName = common_name,
                          speciesNotes = species_notes,
                          occurenceRemarks = Description) %>% 
                          unite(col="higherClassification", c("Huang_Roy_tree_name", "host_clade_sensu_fukami", "complex_robust"), sep=" | ") %>%
                          add_column("namePublishedIn" = "https://doi.org/10.1098/rstb.2014.0010 | https://doi.org/10.1371/journal.pone.0003222 | https://doi.org/10.1038/sdata.2016.17", .after = "higherClassification") %>% # DOI for the relevant publications referenced for higherClassification.
                          separate(physical_sample_location, c("principalInvestigator",                                                               "institutionCode"), sep=",") %>% # Separated physical sample location into principal investigator and institution code, removed institution country.
                          separate(full_taxonomy_string_with_clade, c("phyla", "class", "order"), sep="_") %>%   # Split the phyla, class, order, etc. into their own separate columns.
                          select(-c("taxonomy_string_to_clade", "taxonomy_string_to_family", "taxonomy_string_to_order", "taxonomy_string_to_subclass", "taxonomy_string_to_class")) # Remove the concatenated versions of taxonomy.
### Missing: associatedReferences AND fundingSource, samplingProtocol, previousIdentifications, nomenclaturalCode



Tissue <- Tissue %>% rename(tissueID = SampleID,
                             tissueType = sample_type) # BiologicalMatter also renamed to tissueType?




Diagnostics <- Diagnostics %>% rename(temperatureInDegreesCelsius = temperature,
                                      surfaceTemperatureinDegreesCelsius = surface_temperature,
                                      salinity = salinity, # probably can remove
                                      oxygen = oxygen, # probably can remove
                                      PARinUnits = photosynthetically_active_radiation,
                                      PARMethod = PAR_method,
                                      contactDescription = contact_description,
                                      dominantCover = dominant_cover_2m
                                      coralColonyHeightInCentimeters = colony_height_m,
                                      coralColonyWidth1InCentimeters = colony_width1,
                                      coralColonyWidth2InCentimeters = colony_width2,
                                      coralColonyColorWatch = coral_color,
                                      coralColonyColorWatch2 = coral_color_additional,
                                      coralColonyTissueLoss = binary_tissue_loss,
                                      coralColonyTissueLossPercent = tissue_loss_percent,
                                      diseaseObservedName = disease_name,
                                      diseaseDetected = binary_disease,
                                      diseaseObservedPrevalencePercent = disease_percent,
                                      coralColonyCyanobacteriaContact = cyanobacteria_contact,
                                      coralColonyCyanobacteriaContactPercent = cyanobacteria_percent,
                                      coralColonyCyanobacteriaOvergrown = cyanobacteria_overgrown,
                                      coralColonyTurfContact = binary_turf_contact,
                                      coralColonyTurfContactPercent = turf_contact_percent,
                                      coralColonyTurfOvergrown = turf_overgrown,
                                      coralColonyAlgalContact = binary_algal_contact,
                                      coralColonyAlgalContactsNumber = n_algal_contacts,
                                      coralColonyAlgalContactTypes = algal_contact_types,
                                      coralColonyMacroalgalContact = binary_macroalgal_contact,
                                      coralColonyMacroalgalContactsNumber = n_macroalgal_contacts,
                                      coralColonyMacroalgalContactTypes = macroalgal_contact_types,
                                      coralColonyCCAContact = binary_CCA_contact,
                                      coralColonyCCAContactPercent = cca_contact_percent,
                                      coralColonyCCAOvergrown = cca_overgrown,
                                      coralColonySeagrassContact = seagrass_contact,
                                      coralColonyParrotfishBites = parrotfish_bites,
                                      coralColonyInvertebrateContactTypes = invertebrate_contact_types,
                                      coralColonyCoralContactTypes = coral_contact_types,
                                      coralColonySedimentContact = sediment_contact,
                                      colonyDiameterInCentimeters = Colony_maximum_GCMP_recorded ##??? GCMP column is crossed out in the spreadsheet?
                                      shelfLocation = shelf_location,
                                      visibilityInMeters = visibility,
                                      visibilityMethod = visibility_method)

```

# Configure GEOME

1. Check attributes (column names) of each entity (table) in GEOME, revise definitions to match GCMP definitions, include GCMP term in definition
  a. uncheck any unused fields
  b. move fields into the order they will appear in template spreadsheets
  c. for the Diagnostics entity, identify any terms that are not already entered in GEOME and tell Eric so he can tell Joh to add them.
  d. add "coralColony" to the front of GCMP-specific diagnostics terms
2. Set up lists of controlled vocabularies
3. Set up validation rules for various terms as defined in our working spreadsheet


# Split into Expeditions

There were 11 or 14 GCMP expeditions, and we need to split these concatenated data back into their constituent expeditions.

```{r}
E1 <- df %>% mutate(expedition = str_match(SampleID,"10895\\.(E\\d+)")[,2], .before = sample_name_backup) %>% 
        filter(expedition == "E1")
  

```

# Validate





# Write to Excel Template

You'll want to use the `write.xlsx()` function in the xlsx package to write each table to the `Global Coral Microbiome Project.xlsx` GEOME template in the working directory.